src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extensions.cs

```diff
@@ -0,0 +1,539 @@
+                    }
+
+                    var useSiteInfo = new CompoundUseSiteInfo<AssemblySymbol>(diagnostics, ContainingAssembly);
+                    var current = declaredUnderlyingType;
+                    do
+                    {
+                        if (ReferenceEquals(current.DeclaringCompilation, this.DeclaringCompilation))
+                        {
+                            break;
+                        }
+
+                        current.AddUseSiteInfo(ref useSiteInfo);
+                        current = current.BaseTypeNoUseSiteDiagnostics;
+                    }
+                    while (current is not null);
+
+                    if (!useSiteInfo.Diagnostics.IsNullOrEmpty())
+                    {
+                        var location = FindUnderlyingTypeSyntax(declaredUnderlyingType) ?? Locations[0];
+                        diagnostics.Add(location, useSiteInfo);
+                    }
+
+                    return declaredUnderlyingType;
+                }
+            }
+        }
+
+        private SourceLocation? FindUnderlyingTypeSyntax(TypeSymbol underlyingType)
+        {
+            foreach (var declaration in this.declaration.Declarations)
+            {
+                TypeSyntax? currentSyntax = GetUnderlyingTypeSyntax(declaration);
+                if (currentSyntax != null)
+                {
+                    var baseBinder = this.DeclaringCompilation.GetBinder(currentSyntax);
+                    // Wrap base binder in a location-specific binder that will avoid generic constraint checks.
+                    baseBinder = baseBinder.WithAdditionalFlagsAndContainingMemberOrLambda(BinderFlags.SuppressConstraintChecks, this);
+
+                    var currentType = baseBinder.BindType(currentSyntax, BindingDiagnosticBag.Discarded).Type;
+
+                    if (underlyingType.Equals(currentType, TypeCompareKind.ConsiderEverything))
+                    {
+                        return new SourceLocation(currentSyntax);
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private SingleTypeDeclaration? FirstDeclarationWithExplicitUnderlyingType()
+        {
+            foreach (var singleDeclaration in this.declaration.Declarations)
+            {
+                var underlyingType = GetUnderlyingTypeSyntax(singleDeclaration);
+                if (underlyingType != null)
+                {
+                    return singleDeclaration;
+                }
+            }
+
+            return null;
+        }
+
+        static TypeSyntax? GetUnderlyingTypeSyntax(SingleTypeDeclaration decl)
+        {
+            var syntax = (ExtensionDeclarationSyntax)decl.SyntaxReference.GetSyntax();
+            return syntax.ForUnderlyingType?.UnderlyingType;
+        }
+
+        protected override ImmutableArray<NamedTypeSymbol> BaseExtensionsNoUseSiteDiagnosticsCore
```

Comment:
:memo: I want to point out another choice I made here, to get your feedback. Unlike the corresponding logic in `InterfacesNoUseSiteDiagnostics`, I didn't include a `basesBeingResolved` parameter).
We will likely need to add that when PE symbols enter the picture, but I kept things simpler for now.

```
        internal sealed override ImmutableArray<NamedTypeSymbol> InterfacesNoUseSiteDiagnostics(ConsList<TypeSymbol> basesBeingResolved)
        {
                if (basesBeingResolved != null && basesBeingResolved.ContainsReference(this.OriginalDefinition))
                {
                    return ImmutableArray<NamedTypeSymbol>.Empty;
                }
```